[FLINK-27626] Introduce pre-aggregated merge to table store

HannanKan opened the pull request:
We can introduce richer merge strategies, one of which is already introduced is PartialUpdateMergeFunction, which completes non-NULL fields when merging. This PR introduces support for pre-aggregated merges.

JingsongLi commented on flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/aggregate/FieldAggregator.java:
```diff
@@ -0,0 +1,81 @@
+        return fieldAggregator;
+    }
+
+    abstract void agg(
```
Why not just `Object agg(Object accumulator, Object inputField)`?

JingsongLi commented on flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/aggregate/FieldMinAgg.java:
```diff
@@ -0,0 +1,123 @@
+            min = (mergeField == null ? inField : mergeField);
+        } else {
+            // ordered by type root definition
+            switch (fieldType.getTypeRoot()) {
```
Can we just extract a method `compare` to reuse logical for min and max? Maybe just in `RowDataUtils`.

JingsongLi commented on flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/aggregate/FieldPKAgg.java:
```diff
@@ -0,0 +1,44 @@
+import org.apache.flink.table.types.logical.LogicalType;
+
+/** aggregate by primary keys, so no need to change them except initializing. */
+public class FieldPKAgg extends FieldAggregator {
```
Maybe just use `LAST_VALUE`？

JingsongLi commented on flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/aggregate/RowAggregator.java:
```diff
@@ -0,0 +1,61 @@
+import static org.apache.flink.configuration.ConfigOptions.key;
+
+/** Provide an Aggregator for merge a new row data. */
+public class RowAggregator implements Serializable {
```
Maybe just merge this class into `AggregateMergeFunction`?

JingsongLi reviewed (commented):
Thanks @HannanKan for the contribution.
Left some comments.

JingsongLi reviewed (commented):
Hi @HannanKan , can you add more unit tests to functions? Now there is only itcase.

HannanKan commented:
> Hi @HannanKan , can you add more unit tests to functions? Now there is only itcase.

Good idea! I will do it later.

HannanKan commented:
> Hi @HannanKan , can you add more unit tests to functions? Now there is only itcase.

I have added uint testings for key functions and updated docs. I appreciate if you can review the code again.

HannanKan commented:
@JingsongLi @alpinegizmo @aljoscha @rmetzger  Excuse me. I found that my pr need approve to run github action each time I update branch related to pr. I wonder whether this step could be removed so that we work more efficiently?

JingsongLi reviewed (approved):
Thanks @HannanKan for the update!
Looks good to me!
